crates/autopilot/src/domain/competition/winner_selection.rs

```diff
@@ -54,368 +47,226 @@ use {
-                participant.score().get().0,
-            )
-        });
-        let baseline_scores = compute_baseline_scores(&scores_by_solution);
-        let (fair, unfair) = participants.into_iter().partition_map(|p| {
-            let aggregated_scores = scores_by_solution
-                .get(&SolutionKey {
-                    driver: p.driver().submission_address,
-                    solution_id: p.solution().id(),
-                })
-                .expect("every remaining participant has an entry");
-            // only keep solutions where each order execution is at least as good as
-            // the baseline solution.
-            // we only filter out unfair solutions with more than one token pair,
-            // to avoid reference scores set to 0.
-            // see https://github.com/fhenneke/comb_auctions/issues/2
-            if aggregated_scores.len() == 1
-                || aggregated_scores.iter().all(|(pair, score)| {
-                    baseline_scores
-                        .get(pair)
-                        .is_none_or(|baseline| score >= baseline)
-                })
-            {
-                Either::Left(p)
-            } else {
-                Either::Right(p)
-            }
-        });
-        PartitionedSolutions {
-            kept: fair,
-            discarded: unfair,
+        let mut ranked = Vec::with_capacity(ws_ranking.ranked.len());
+        for ranked_solution in ws_ranking.ranked {
+            let key = SolutionKey::from(&ranked_solution);
+            let participant = participant_by_key
+                .remove(&key)
+                .expect("every ranked solution has a matching participant");
+            let score = ranked_solution.score();
+            ranked.push(
+                participant
+                    .with_score(Score(eth::Ether(score)))
+                    .with_rank(ranked_solution.state().rank_type),
+            );
         }
-    }
 
-    /// Picks winners and sorts all solutions where winners come before
-    /// non-winners and higher scores come before lower scores.
-    fn mark_winners(&self, participants: Vec<Participant<Scored>>) -> Vec<Participant<Ranked>> {
-        let winner_indexes = self.pick_winners(participants.iter().map(|p| p.solution()));
-        participants
-            .into_iter()
-            .enumerate()
-            .map(|(index, participant)| {
-                let rank = match winner_indexes.contains(&index) {
-                    true => RankType::Winner,
-                    false => RankType::NonWinner,
-                };
-                participant.with_rank(rank)
-            })
-            .collect()
+        Ranking {
+            filtered_out,
+            ranked,
+        }
     }
 
     /// Computes the reference scores which are used to compute
     /// rewards for the winning solvers.
     pub fn compute_reference_scores(&self, ranking: &Ranking) -> HashMap<eth::Address, Score> {
```

Comment:
Since we now have to do another conversion it might make more sense now to already compute the reference scores as part of `arbitrate` and return it inside `Ranking` somehow?